libs/coroutines/include/hpx/coroutines/coroutine.hpp

```diff
@@ -34,49 +34,46 @@
 #include <hpx/config.hpp>
 
 #include <hpx/assertion.hpp>
-#include <hpx/runtime/threads/coroutines/coroutine_fwd.hpp>
-#include <hpx/runtime/threads/coroutines/detail/coroutine_accessor.hpp>
-#include <hpx/runtime/threads/coroutines/detail/coroutine_impl.hpp>
-#include <hpx/runtime/threads/coroutines/detail/coroutine_self.hpp>
-#include <hpx/runtime/threads/thread_enums.hpp>
-#include <hpx/runtime/threads/thread_id_type.hpp>
-#if defined(HPX_HAVE_APEX)
-#include <hpx/util/apex.hpp>
-#endif
+#include <hpx/coroutines/coroutine_fwd.hpp>
+#include <hpx/coroutines/detail/coroutine_accessor.hpp>
+#include <hpx/coroutines/detail/coroutine_impl.hpp>
+#include <hpx/coroutines/detail/coroutine_self.hpp>
+#include <hpx/coroutines/thread_enums.hpp>
+#include <hpx/coroutines/thread_id_type.hpp>
 
 #include <cstddef>
 #include <cstdint>
 #include <limits>
 #include <utility>
 
-namespace hpx { namespace threads { namespace coroutines
-{
+namespace hpx { namespace threads { namespace coroutines {
     /////////////////////////////////////////////////////////////////////////////
+    template <typename ThreadData>
```

Comment:
Would it be more appropriate to rely on an abstract base class instead of making the coroutines be templatized on the `thread_data` type? We will need to type-erase the `thread_data` anyways in order to add stack-less HPX threads later on.

Reply:
The reason I prefer this is that this way the coroutines module doesn't need to know *anything* about thread data (which is appropriate as the only use in the module is to store a pointer to thread data, and only return it to the outside world). However, if getting stackless threads in later will be easier with an abstract base class, then it's an option of course. How much of the current `thread_data` interface do you think would be required in `thread_data_base`? `thread_id` could of course be templated *and* there could be an abstract base class in the same module as `thread_data`?

Reply:
The fact that `thread_id` actually holds the pointer to the thread data is an implementation detail that shouldn't ever have leaked into the code base. But you know how these things happen... For this reason, I think `thread_id` does not need to be templated on the thread data. Conceptually it's just an opaque id that somehow identifies the thread instance. 

It might be sufficient to store a `void*` or even a `std::size_t` in the `thread_id` and add a (friend) helper function `thread_data* get_thread_data(thread_id id);` that knows how to wield the magic.

For the abstract base class, I think we would need it in any case if we decided to add stack-less threads. I don't see another way for the scheduler to be agnostic of the thread type to handle. We will need some type of type erasure there for sure.